### PR TITLE
Support for patch file

### PR DESCRIPTION
**Pull request by jgeudens**
Implementation for support of a patch file after fetching (#22)

- [x] The documentation in ~~manifest.py~~ project.py explaining this new attribute.
- [x] Feature tests
- ~~Unit tests~~
- [x] Add patch: to the schema for validation to succeed.
- [x] Add any external lib license to legal.rst (and make sure it is MIT compatible)

### CODE REVIEW

**Comment by jgeudens** [comment]
Current state:

* Patch option has been added to manifest (with unit test)
* Basic feature test has been added
* Added function to apply path

At the moment, the init of the patch-ng library fails with the patch from from the feature file.

**Comment by ben-edna** [comment]
Maybe also consider a feature test, if the patch fails.
Does the patch-ng lib have a function call to check of the patch can be applied? If so, call that, and if the patch cannot be applied, warn the user and ignore the patch

**Comment by jgeudens** [comment]
I have added feature tests (git and svn, success and fail) and documentation (legal.rst and more). I'm not sure how to unit test this feature.

Can you perform an initial review?

File: dfetch/project/vcs.py [hunk 1]
```diff
@@ -101,6 +102,10 @@ def update(self, force: bool = False) -> None:
 
         actually_fetched = self._fetch_impl(to_fetch)
         self._log_project(f"Fetched {actually_fetched}")
+
+        if self.__project.patch and os.path.exists(self.__project.patch):
```
Comment (ben-edna) [inline dfetch/project/vcs.py:106]: Could you add this to the uml diagram in `docs/static/uml`?

File: dfetch/project/vcs.py [hunk 2]
```diff
@@ -109,6 +114,20 @@ def update(self, force: bool = False) -> None:
...
+    def apply_patch(self) -> None:
+        """ Apply the specified patch to the destination """
+
+        patch_set = fromfile(self.__project.patch)
+
+        if patch_set:
+            if patch_set.apply(0, root=self.__project.destination, fuzz=True):
+                self._log_project(f"Applied path \"{self.__project.patch}\"")
+            else:
+                self._log_project(f"Applying path \"{self.__project.patch}\" failed")
```
Comment (ben-edna) [inline dfetch/project/vcs.py:127]: maybe we should propagate this up as a runtime-error, then the user sees red texts and gets a non-zero return code.

```python
raise RuntimeError("Applying patch failed")
```

File: features/patch-after-fetch-svn.feature [hunk 3]
```diff
@@ -0,0 +1,80 @@
+Feature: Patch after fetching from git repo
+
+    Sometimes a patch needs to be applied after fetching. *DFetch* makes it
+    possible to specify a patch file.
+
+    @wip
```
Comment (ben-edna) [inline features/patch-after-fetch-svn.feature:6]: There are some stray `@wip` tags in this file 😉 , probably a development artifact

File: features/steps/generic_steps.py [hunk 4]
```diff
@@ -15,6 +15,19 @@
 dfetch_title = re.compile(r"Dfetch \(\d+.\d+.\d+\)")
 
 
+def check_file(path, content):
```
Comment (ben-edna) [inline features/steps/generic_steps.py:18]: Nice!

File: features/steps/generic_steps.py [hunk 5]
```diff
@@ -58,6 +71,12 @@ def list_dir(path):
     return result
 
 
+@given("the patch file '{name}'")
+def step_impl(context, name):
+
```
Comment (ben-edna) [inline features/steps/generic_steps.py:76]: Maybe `black` the steps folder as well.

File: features/patch-after-fetch-svn.feature [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            ===================================================================
+            --- README.md	(revision 2)
+            +++ README.md	(revision 3)
+            @@ -1,2 +1,2 @@
+             # Test-repo
+            -A test repo for testing dfetch.
+            +A test repo for testing patch.
+            """
+        When I run "dfetch update"
+        Then the resulting file after patching should be 'ext/test-repo-tag/README.md'
```
Comment (ben-edna) [inline features/patch-after-fetch-svn.feature:36]: How do you feel about a more concise sentence:
```suggestion
        Then the patched 'ext/test-repo-tag/README.md' is
```

**Review by ben-edna — CHANGES_REQUESTED** [review]
Thanks for the PR!

Did you run `pre-commit run --all-files` ? This might show you some additional items to fix (small hint, don't touch the github workflow files e.g. `test.yml`).

Also don't forget to update the `CHANGELOG.rst`

File: dfetch/project/vcs.py [hunk 7]
```diff
@@ -10,6 +10,7 @@
 from dfetch.project.version import Version
 from dfetch.util.util import hash_directory, safe_rm
 from dfetch.util.versions import latest_tag_from_list
+from patch_ng import fromfile
```
[the remark this answers is not in the document]
Comment (jgeudens) [reply dfetch/project/vcs.py]: Yes, I have ran pre-commit now.

File: dfetch/project/vcs.py (hunk 1 above)
Comment (jgeudens) [reply dfetch/project/vcs.py:107]: Done

File: dfetch/project/vcs.py (hunk 2 above)
Comment (jgeudens) [reply dfetch/project/vcs.py]: Done,

should the feature file check return code?

File: features/patch-after-fetch-svn.feature (hunk 3 above)
Comment (jgeudens) [reply features/patch-after-fetch-svn.feature]: Removed

File: features/steps/generic_steps.py (hunk 4 above)
Comment (jgeudens) [reply features/steps/generic_steps.py:18]: thanks

File: features/steps/generic_steps.py (hunk 5 above)
Comment (jgeudens) [reply features/steps/generic_steps.py:76]: Yes, I have ran pre-commit now.

File: features/patch-after-fetch-svn.feature (hunk 6 above)
Comment (jgeudens) [reply features/patch-after-fetch-svn.feature]: Made more concise

**Comment by jgeudens** [comment]
I think I have addressed all your remarks. Can you review again?

File: dfetch/project/vcs.py (hunk 2 above)
Comment (ben-edna) [reply dfetch/project/vcs.py]: No, not concern of this feature in my opinion. What do you think?

Comment (jgeudens) [reply dfetch/project/vcs.py]: Other feature files also don't check return code, so I think it is ok.

**Review by ben-edna — APPROVED** [review]

**Comment by ben-edna** [comment]
There are some doc issues still

**Comment by jgeudens** [comment]
> There are some doc issues still

I fixed the duplicate "here" id, but CI still fails. It appears something to do with the API token.

**Comment by ben-edna** [comment]
> 
> 
> > There are some doc issues still
> 
> I fixed the duplicate "here" id, but CI still fails. It appears something to do with the API token.

I'll fix it and merge, thanks

**Pull request merged**